fuel-core/src/executor.rs

```diff
@@ -897,67 +903,175 @@ impl Executor {
         Ok(min_fee)
     }
 
-    /// In production mode, lookup and set the proper utxo ids for contract inputs
-    /// In validation mode, verify the proposed utxo ids on contract inputs match the expected values.
-    fn compute_contract_input_utxo_ids(
+    /// Computes all zeroed or variable inputs.
+    /// In production mode, updates the inputs with computed values.
+    /// In validation mode, compares the inputs with computed inputs.
+    fn compute_inputs<Tx>(
         &self,
-        tx: &mut ExecutionTypes<&mut Transaction, &Transaction>,
-        db: &Database,
-    ) -> Result<(), Error> {
-        let expected_utxo_id = |contract_id| {
-            let maybe_utxo_id = db.storage::<ContractsLatestUtxo>().get(contract_id)?;
+        tx: ExecutionTypes<&mut Tx, &Tx>,
+        db: &mut Database,
+    ) -> Result<(), Error>
+    where
+        Tx: ExecutableTransaction,
+    {
+        let expected_utxo_id = |contract: &ContractRef<&mut Database>| {
+            let maybe_utxo_id = contract.utxo()?.map(|utxo| utxo.into_owned());
             let expected_utxo_id = if self.config.utxo_validation {
                 maybe_utxo_id
-                    .ok_or(Error::ContractUtxoMissing(*contract_id))?
-                    .into_owned()
+                    .ok_or_else(|| Error::ContractUtxoMissing(*contract.contract_id()))?
             } else {
-                maybe_utxo_id.unwrap_or_default().into_owned()
+                maybe_utxo_id.unwrap_or_default()
             };
             Result::<_, Error>::Ok(expected_utxo_id)
         };
```

Comment:
Instead of a lambda, can `expected_utxo_id` be replaced by a getter on the `ContractRef` interface? If that is practical, I think it would make the readability a bit better and consistent:

```rust
let mut contract = ContractRef::new(&mut *db, contract_id);
*utxo_id = contract.utxo_id()?;
*balance_root = contract.balance_root()?;
*state_root = contract.state_root()?;
```

Reply:
I tried to avoid any business logic inside of those wrappers=)

Reply:
Good point. I am not strongly opinionated either way. As usual with software, there are pros and cons to either approach!